### ISSUE TITLE
feat: semver locking

### ISSUE THREAD

**Issue by cafkafk** [cafkafk/rime#10]
It would be cool if we could lock to a specific semver major, minor, or patch version. Further, we might consider if it's possible to have logic, such as great than version x.y.z, less than version a.b.c

**Comment by algernon** [comment]
I think the major problem with this is that the only version Rime has easy access to are tags, and different projects use different tag naming methods: some just use the version as-is, some prefix it with `v`, some prefix it with the project name. I guess Rime could do a best effort at finding a version within a tag name, then use the `semver` crate and query params to do some matching?

Something along these lines, perhaps:

```rust
use semver::{Version, VersionReq};

let req = VersionReq::parse(query['semver'])?;
let tag = repo.tags.find(|&tag| {
  let version = Version::parse(pick_version_from_tag(tag)).unwrap();
  req.matches(&version)
});

// redirect to /api/v1/..../v/<tag>
```

**Comment by cafkafk** [comment]
One issue I'm not sure if we can overcome is we (seemingly at least) need to end the url the user requests with `.tar.gz` for nix, and that may or may not make query parameters harder.

But yea, I was thinking some sort of best effort thing, first we attempt to go for ones that prefix with `v`, then we see if there are any tags that match semver without `v`.

I hadn't thought of doing `/semver/:major/:minor/:patch` but that is a great approach, not sure how we do `-rc` `-alpha` and such suffixes thou.

**Comment by algernon** [comment]
> One issue I'm not sure if we can overcome is we (seemingly at least) need to end the url the user requests with `.tar.gz` for nix, and that may or may not make query parameters harder.

Quick testing suggests that query parameters are fine, Nix handles them appropriately (ie, passes them on, but otherwise treats the URL as ending in `.tar.gz`):

```
❯ nix run 'https://git.madhouse-project.org/algernon/riemann-c-client/archive/main.tar.gz?foo=bar' -- --version
riemann-c-client 2.1.2-snapshot
```

Query params also seem to work in `flake.nix` inputs, so I think we're fine!

> I hadn't thought of doing `/semver/:major/:minor/:patch` but that is a great approach, not sure how we do `-rc` `-alpha` and such suffixes thou.

Maybe a `pre` and `build` query argument? Don't think we can bake that into the path, because both are optional, and either can be specified independently of the other. Perhaps an `/:extra` path could work, and it'd be used for both the `-rc`, `-alpha`, etc suffixes *and* build metadata? So, `/semver/1/0/0/-rc1+foo`?

Or, wait a minute... we don't even need to do `/:major/:minor/:patch`, we can just `/:major.:minor.:patch`, too, can't we? Or even more simply: `/semver/:version`, and let the `semver` crate parse the version as a version requirement. That'd allow semver pinning the same way Cargo does when specifying a version. For more elaborate requirements, `/semver?req=....` would do the trick.

**Comment by cafkafk** [comment]
`/semver/:version` does seem simple, also TIL build metadata exists

**Comment by algernon** [comment]
I've been pondering about how to approach implementing this, in a way that'd benefit #11 too, and came up with the following idea: rather than providing a `get_latest_tag` -like function in some forges, they'd provide `get_releases`, which would get a set number of releases, and turn them into a common format:

```rust
struct ForgeRelease {
  tag_name: String,
  prerelease: bool,
}
```

Then, we could just pick the topmost one if pre-releases are enabled, and the first non-pre-release otherwise for the `:user/:repo.tar.gz` routes. And for semver, a shared code could match against those in the list.

**Comment by cafkafk** [comment]
I think such an approach makes sense, and could centralize a fair amount of code :+1:

**Comment by algernon** [comment]
To summarize things so far, in light of what I ended up with in #20:

`/api/v1/forgejo/:host/:user/:repo/semver/:version` and `/api/v1/forgejo/:host/:user/:repo/s/:version` would be the API endpoints I'd propose for this feature (and the same thing for gitlab, and github too). These two would cover the simple case, which corresponds to `crate = { version = "x.y.z" }` in `Cargo.toml`. Under the hood, Rime would parse `:version` into a `semver::VersionReq`.

Would someone require more control, the `.../semver` and `../s` endpoints and a `?version=[version_req]` query parameter would be provided. The `version_req` string there would be parsed into a `semver::VersionReq` under the hood.

Both endpoints would delegate the matching to a shared function that iterates over the releases in a `ForgeReleases` argument, and matches each version against the `VersionReq`. Said function would also try to discover which part of the tag name is an actual version (likely delegating that to another function).

**Comment by algernon** [comment]
Come to think of it.... `/api/v1/forgejo/:host/:user/:repo.tar.gz` could be a special case of `/api/v1/forgejo/:host/:user/:repo/semver/*.tar.gz`. Except that the `repo.tar.gz` path might not conform to semver, so we can't use this kind of redirect.

Thought I'll write that down, because I'm sure I will forget it while I sleep.

**Comment by algernon** [comment]
Okay, this is actually incredibly useful:

```
❯ FORGE="forgejo/git.madhouse-project.org"; alias loc="grep location"
❯ curl -si "http://127.0.0.1:3000/v1/$FORGE/algernon/rime-testing/s/*.tar.gz" | loc
location: /v1/forgejo/git.madhouse-project.org/algernon/rime-testing/v/2.0.0.tar.gz
❯ curl -si "http://127.0.0.1:3000/v1/$FORGE/algernon/rime-testing/s/1.*.tar.gz" | loc
location: /v1/forgejo/git.madhouse-project.org/algernon/rime-testing/v/1.2.3.tar.gz
```

(Proof of concept [here](https://github.com/algernon/rime/commit/bc5694cbec3d79bb4a2c5ea4ffd903db0c88e765))

Mind you, this approach would result in an uncomfortable amount of copypasta. I have an idea how to hack most of it away, but will need to do a bit of experimenting first. I'll finish the semver locking PR before that, however, regardless of how much duplicated code I end up writing as part of it.

**Comment by algernon** [comment]
That `*.tar.gz` part is :ok_hand:. So much so, that I'd love to have something similar for `/v/` and `/b/` too. Not semver-based, of course, but glob-based maybe? Could be useful to pin `nixpkgs` via Rime: `/v1/github/NixOS/nixpkgs/b/nixos-??.??` would let me pin to the latest nixos, whatever it may be, rather than any specific version...

But talking about nixpkgs, here's another fun problem: nixpkgs does not use releases. They have tags and branches. So `/v1/github/nixos/nixpkgs.tar.gz` currently doesn't work (no releases), and neither would the semver endpoint. Yet, handling nixpkgs through Rime would still provide benefits (such as being able to pin to `nixpkgs/nixos-??.??`). Something to ponder about at a later time.

**Comment by cafkafk** [comment]
> But talking about nixpkgs, here's another fun problem: nixpkgs does not use releases. They have tags and branches. So `/v1/github/nixos/nixpkgs.tar.gz` currently doesn't work (no releases), and neither would the semver endpoint. Yet, handling nixpkgs through Rime would still provide benefits (such as being able to pin to `nixpkgs/nixos-??.??`). Something to ponder about at a later time.

I think we talked about downgrading an endpoint if there were no releases to just do semver based stuff, I wonder if such a thing would make sense here (obvs not semver based). I am a bit apprehensive of the easy solution of just making a special case for nixpkgs, because I honestly kinda feel like we shouldn't enable them in being unique and quirky. At the same time, they're just such a fundamental part of the nix ecosystem that we might have to do it anyways

I wonder if we can make it somewhat generic thou. But I've not explored how to deal with nixpkgs at all yet just wanted to write my thoughts down.  This would obviously be super valuable thou.

**Comment by algernon** [comment]
Oh, definitely no special casing for nixpkgs. Enabling globbing for `/b/` and `/v/` in general, would cover nixpkgs (well, `/b/` would, since `/v/` requires releases, which nixpkgs doesn't have), and likely many others that use more complicated branching strategies.

I'll open an issue about it once I had a bit more time thinking, and once I found examples other than nixpkgs, where such a feature would be beneficial.

**Comment by algernon** [comment]
As it happens, this endpoint is super useful for use cases outside of Nix! If I just want the a tarball of any project that follows semver, and is on a supported forge, one where the version matches a particular semver version requirement, I can construct such a query via Rime!

**Comment by algernon** [comment]
> Oh, definitely no special casing for nixpkgs. Enabling globbing for `/b/` and `/v/` in general, would cover nixpkgs (well, `/b/` would, since `/v/` requires releases, which nixpkgs doesn't have), and likely many others that use more complicated branching strategies.
> 
> I'll open an issue about it once I had a bit more time thinking, and once I found examples other than nixpkgs, where such a feature would be beneficial.

I've been thinking about this during the day, and I see problems with implementing globbing for `/b/` and `/v/`: we'd need to do a whole lot more talking to remote forges. So I'd rather improve the `/semver/:version` endpoint, and teach it to match against not just releases, but branches or tags too. We'd still need a few more helpers to query remote branches and non-release tags, but it would be constrained to the `/semver` route, which is already on the expensive side.

The core `/semver/:version` route would remain the same: query the remote releases, match against the spec. But I'd allow three more query parameters (on top of the existing `version` param I implemented in #31):

- `match-against=[releases|tags|branches]`: this would select what to match against.
- `strip-prefix=<string>`: this would simply strip the given prefix from the branch name **and** from the version requirement spec, before trying to parse the rest of it as a semantic version.

Thus: `/v1/github/nixos/nixpkgs/semver/nixos-*.tar.gz?match-against=branches&strip-prefix=nixos-` would match the latest `nixpkgs/nixos-xx.yy` branch (because `nixos-xx.yy-<something else>` does not parse as semver, luckily).

Alternatively, we could make the prefix finding a bit more intelligent: it'd try to parse the whole thing as semver, then strip the first char, try again, and repeat until the parse succeeds, or we reach the end of the string (doing this after the existing `v` and `{repo}-` strips, so we have shortcuts).

Doing this would allow me to use `/v1/github/nixos/nixpkgs/semver/nixos-*.tar.gz?match-against=branches`. Or perhaps the semver route could automatically fall back to branches, then to tags, further simplyfing the request to `/v1/github/nixos/nixpkgs/semver/nixos-*.tar.gz`. In which case, it would first strip away `nixos-`, and use `*` as the semver version req. Then, it would list the releases, find none. Then it would list the branches, and find `nixos-*` branches (because branch and tag finding would use the original path as the glob), then remove the `nixos-` prefix from all of them (like it removed for the version req), and match the remaining version number against the version req.

I think this would be pretty darn sweet, and flexible.

However, I'm yet to find anything *other than* nixpkgs that would benefit from this complexity. Even if the code would *technically* not be a special case for nixpkgs, if there's nothing else benefiting from this, then it practically is a special case.

**Comment by cafkafk** [comment]
> However, I'm yet to find anything other than nixpkgs that would benefit from this complexity. Even if the code would technically not be a special case for nixpkgs, if there's nothing else benefiting from this, then it practically is a special case.

I just realized now that there actually is, you can have tagged pre-releases in semver, e.g. `1.6.1-pre.2`, but even something like `1.6.1-pre.2.4.5.5.5.5.5` is apparently valid, and in the case of those being used for branch names (not something I've ever observed in the wild), this feature might also be useful... ish, but that does *technically* make this not a special case.
